File: shap_analysis.R
# 1. 加载包
library(fastshap)
library(shapviz)
library(dplyr)
library(ggplot2)

# 2. 读取数据
heart_data <- read.csv("data/heart_clean.csv")

# 3. 将分类变量转为因子（与 Python 独热编码对应）
# 注意：这些变量在原始数据中虽然是数字编码，但本质是分类
heart_data$sex <- as.factor(heart_data$sex)
heart_data$cp <- as.factor(heart_data$cp)
heart_data$fbs <- as.factor(heart_data$fbs)
heart_data$restecg <- as.factor(heart_data$restecg)
heart_data$exang <- as.factor(heart_data$exang)
heart_data$slope <- as.factor(heart_data$slope)
heart_data$thal <- as.factor(heart_data$thal)

# 目标变量转为因子（二分类）
heart_data$target <- as.factor(heart_data$target)

# 4. 拟合逻辑回归（glm 会自动创建虚拟变量，参考水平为第一个因子水平）
model <- glm(target ~ ., data = heart_data, family = binomial())

# 查看模型摘要（可选）
summary(model)

# 5. 定义预测包装器（输出概率）
pred_wrapper <- function(object, newdata) {
  predict(object, newdata, type = "response")
}

# 6. 准备解释数据（去除 target 列）
X_explain <- heart_data %>% dplyr::select(-target)

# 7. 计算 SHAP 值（使用所有数据，若较慢可采样前200行）
set.seed(123)
shap_values <- fastshap::explain(
  model,
  X = X_explain,
  pred_wrapper = pred_wrapper,
  nsim = 50,
  adjust = TRUE
)

# 8. 转换为 shapviz 对象
shp <- shapviz(shap_values, X = X_explain)

# 9. 保存 SHAP summary plot (beeswarm)
png("output/shap_summary.png", width = 8, height = 6, units = "in", res = 150)
sv_importance(shp, kind = "beeswarm") +
  labs(title = "SHAP Summary Plot - MACE Prediction (Corrected)")
dev.off()

# 10. 保存 SHAP bar plot (特征重要性)
png("output/shap_bar.png", width = 8, height = 5, units = "in", res = 150)
sv_importance(shp, kind = "bar") +
  labs(title = "SHAP Feature Importance (Bar Plot)")
dev.off()

# 11. 打印完成信息
print("修正后的 SHAP 图已保存至 output/shap_summary.png 和 output/shap_bar.png")

File: shap_analysis.py
import shap
import pandas as pd
import joblib
import matplotlib.pyplot as plt
import os

# 加载数据与模型
df = pd.read_csv("data/heart_clean.csv")
cat_cols = ["sex", "cp", "fbs", "restecg", "exang", "slope", "thal"]
df_encoded = pd.get_dummies(df, columns=cat_cols, drop_first=True)
X = df_encoded.drop("target", axis=1)
y = df_encoded["target"]

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

lr = joblib.load("models/logistic_model.pkl")

# 使用 KernelExplainer（适用于任何模型，但较慢）或 LinearExplainer（更快）
# 尝试使用 shap.explainers.Linear（如果可用）
try:
    # 新版 shap 路径
    explainer = shap.explainers.Linear(lr, X_train)
except AttributeError:
    # 旧版或兼容方式
    explainer = shap.LinearExplainer(lr, X_train, feature_dependence="independent")

shap_values = explainer.shap_values(X_test)

# 如果是二分类，shap_values 可能是一个列表，取正类
if isinstance(shap_values, list):
    shap_values = shap_values[1]

# SHAP summary plot
shap.summary_plot(shap_values, X_test, show=False)
plt.tight_layout()
plt.savefig("output/shap_summary.png", dpi=150, bbox_inches="tight")
plt.close()

# SHAP bar plot
shap.summary_plot(shap_values, X_test, plot_type="bar", show=False)
plt.tight_layout()
plt.savefig("output/shap_bar.png", dpi=150, bbox_inches="tight")
plt.close()

print("SHAP 图已保存至 output/shap_summary.png 和 output/shap_bar.png")